dingo/gw/noise/synthetic_noise/asd_sampling.py

```diff
@@ -0,0 +1,112 @@
+import copy
+
+import numpy as np
+from scipy import stats
+from dingo.gw.noise.synthetic_noise.asd_parameterization import fit_broadband_noise
+from dingo.gw.noise.asd_dataset import ASDDataset
+from dingo.gw.noise.synthetic_noise.utils import (
+    get_index_for_elem,
+)
+
+
+class KDE:
+    def __init__(self, parameter_dict, sampling_settings):
+
+        self.parameter_dicts = parameter_dict
+        self.settings = sampling_settings
+
+        detectors = parameter_dict.keys()
+        self.spectral_kde = {det: [] for det in detectors}
+        self.broadband_kde = {det: [] for det in detectors}
+
+    def fit(self, weights=None):
+
+        for det, param_dict in self.parameter_dicts.items():
+            y_values = param_dict["y_values"]
+            x_positions = param_dict["x_positions"]
+            spectral_features = param_dict["spectral_features"]
+
+            for i in range(spectral_features.shape[1]):
+                try:
+                    spectral_kde = stats.gaussian_kde(
+                        spectral_features[:, i, :].T,
+                        bw_method=float(self.settings["bandwidth_spectral"]),
+                        weights=weights,
+                    )
+                except np.linalg.LinAlgError:
+                    print(
+                        "Warning: Singular Matrix encountered in spectral KDE. Adding small Gaussian noise..."
+                    )
+                    perturbed_features = spectral_features[:, i, :] + np.random.normal(
+                        0, 1.0e-4, size=spectral_features[:, i, :].shape
+                    )
+                    spectral_kde = stats.gaussian_kde(
+                        perturbed_features.T,
+                        bw_method=float(self.settings["bandwidth_spectral"]),
+                        weights=weights,
+                    )
+                self.spectral_kde[det].append(spectral_kde)
+
+            split_indices = [0, len(x_positions)]
+            split_indices += [
+                get_index_for_elem(x_positions, f) + 1
+                for f in self.settings["split_frequencies"]
+            ]
+            split_indices = sorted(split_indices)
+
+            for i in range(len(split_indices) - 1):
+                vals = y_values[:, split_indices[i]: split_indices[i + 1]].T
+                kde_vals = stats.gaussian_kde(
+                    vals, bw_method=float(self.settings["bandwidth_spline"])
+                )
+                self.broadband_kde[det].append(kde_vals)
+
+    def sample(self, rescaling_ys):
```

Comment:
Should `rescaling_ys` have a default argument of `None`, for the case where we do not want to rescale the sampled PSDs? The code seems to accept `None`, but it should also probably be the default.

I think it would be useful for this method to also have a docstring.

Reply:
Yeah, I totally agree